## Setup Configuration and Read Credentials

In [4]:
#!pip3 install python-dotenv
#!pip3 install sqlalchemy
#!pip3 install psycopg2-binary

In [5]:
#from dotenv import load_dotenv


In [6]:
#load_dotenv('.env')

In [7]:
#import os
#os.getenv('db')

# Proyek SQL - Aplikasi Pecinta Buku


#  Gambaran Umum

Virus corona yang kehadirannya mengejutkan seluruh dunia telah mengubah rutinitas harian semua orang. Kini, warga kota tidak lagi menghabiskan waktu luangnya di luar rumah seperti pergi ke kafe dan mal. 

Mereka lebih sering berada di rumah, menghabiskan waktunya dengan membaca buku. Hal ini pun mendorong perusahaan startup untuk mengembangkan aplikasi baru bagi para pecinta buku.

Anda telah diberi sebuah database dari salah satu perusahaan yang bersaing dalam industri ini. Database tersebut berisi data tentang buku, penerbit, penulis, serta rating dan ulasan pelanggan atas buku terkait. 
Informasi ini akan digunakan dalam membuat penawaran harga untuk sebuah produk baru.

# Tujuan studi

Secara khusus, beberapa tujuan utama dari studi ini meliputi:

1. Mengidentifikasi Jumlah Buku Terbit setelah 1 Januari 2000:
   Hal ini dapat memberikan gambaran tentang jumlah buku yang tersedia dalam periode waktu tertentu.

2. Menganalisis Ulasan dan Peringkat: 
   Menghitung jumlah ulasan pengguna dan rata-rata peringkat untuk setiap buku. Ini dapat membantu dalam mengevaluasi  popularitas dan kualitas buku-buku tertentu.

3. Mengidentifikasi Penerbit Terkemuka: 
   Mengidentifikasi penerbit yang telah merilis jumlah buku terbanyak dengan lebih dari 50 halaman. Ini membantu dalam mengenali penerbit yang memiliki kontribusi besar dalam industri buku.

3. Mencari Penulis dengan Peringkat Tertinggi: 
   Mengidentifikasi penulis dengan rata-rata peringkat buku tertinggi, dengan syarat bahwa hanya buku dengan setidaknya 50 ulasan yang akan dipertimbangkan. Ini dapat membantu dalam mengevaluasi reputasi penulis.

4. Menganalisis Jumlah Ulasan Teks: 
   Menemukan rata-rata jumlah ulasan teks di antara pengguna yang memberi peringkat lebih dari 50 buku. Ini dapat memberikan wawasan tentang seberapa aktif pengguna dalam memberikan ulasan.

## Read Connection From PSQL DB

In [8]:
# impor library yang diperlukan
import pandas as pd
from sqlalchemy import create_engine


In [9]:

# Konfigurasi koneksi ke database
db_config = {'user'  : '            ',         # nama pengguna
             'pwd'   : '            ',          # kata sandi
             'host'  : '            ',         # nama host
             'port'  : '            ',    # port koneksi
             'db'    : '            '}           # nama database

# Membentuk string koneksi
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(
    db_config['user'],
    db_config['pwd'],
    db_config['host'],
    db_config['port'],
    db_config['db'])

# Membuat koneksi ke database
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Check Table Name

In [10]:
# Mengecek tabel-tabel yang ada dalam database
query = "SELECT table_name FROM information_schema.tables WHERE table_schema = 'public';"

# Menjalankan kueri SQL
tables = pd.read_sql(query, engine)

# Menampilkan daftar tabel
print("Tabel-tabel dalam database:")
print(tables)


Tabel-tabel dalam database:
           table_name
0  advertisment_costs
1             authors
2               books
3              orders
4          publishers
5             ratings
6             reviews
7              visits


In [11]:
# akan mengatur batas tampilan hingga 50 baris.
pd.set_option('display.max_rows', 50)


In [12]:
# mengambil informasi tentang semua tabel yang ada di database
query = """SELECT * FROM pg_catalog.pg_tables;"""

In [13]:
# menjelajahi informasi tentang tabel-tabel dalam database
df = pd.io.sql.read_sql(query, con = engine)

In [14]:
#  mengambil tabel-tabel yang hanya berada dalam skema 'public' di database
df[df['schemaname'] == 'public']

schemaname           tablename         tableowner tablespace  hasindexes  \
66     public              author  praktikum_student       None        True   
67     public              second  praktikum_student       None       False   
68     public            second_b  praktikum_student       None       False   
69     public  advertisment_costs    praktikum_admin       None        True   
70     public             authors    praktikum_admin       None        True   
71     public               books    praktikum_admin       None        True   
72     public              orders    praktikum_admin       None        True   
73     public          publishers    praktikum_admin       None        True   
74     public             ratings    praktikum_admin       None        True   
75     public             reviews    praktikum_admin       None        True   
76     public              visits    praktikum_admin       None        True   

    hasrules  hastriggers  rowsecurity  
66     False        False        False  
67     False        False        False  
68     False        False        False  
69     False        False        False  
70     False         True        False  
71     False         True        False  
72     False        False        False  
73     False         True        False  
74     False         True        False  
75     False         True        False  
76     False        False        False

Tabel-tabel dalam database: table_name

1 authors 2 books 4 publishers 5 ratings 6 reviews

0 advertisment_costs 3 orders 7 visits

In [9]:

# Konfigurasi koneksi ke database
db_config = {'user'  : 'practicum_student',         # nama pengguna
             'pwd'   : 's65BlTKV3faNIGhmvJVzOqhs',          # kata sandi
             'host'  : 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',         # nama host
             'port'  : '6432',    # port koneksi
             'db'    : 'data-analyst-final-project-db'}           # nama database

# Membentuk string koneksi
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(
    db_config['user'],
    db_config['pwd'],
    db_config['host'],
    db_config['port'],
    db_config['db'])

# Membuat koneksi ke database
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Explore Table

### Explore table publishers

In [15]:
# Menampilkan data di table books
query_publishers = """select * from public.publishers;"""
data_publishers = pd.io.sql.read_sql(query_publishers, con=engine)

data_publishers

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

In [16]:
data_publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


In [17]:
# Menyusun kueri SQL untuk mendapatkan nama kolom dari tabel publishers
query_publishers = """ 
SELECT column_name 
FROM information_schema.columns
WHERE table_name = 'publishers';"""

# Membaca nama kolom ke dalam dataframe
data_publishers = pd.io.sql.read_sql(query_publishers, con=engine)
print("Nama kolom dalam tabel Publishers:")
data_publishers


Nama kolom dalam tabel Publishers:


column_name
0  publisher_id
1     publisher

### Explore table authors

In [18]:
query_authors = """SELECT * FROM public.authors;"""
data_authors = pd.io.sql.read_sql(query_authors, con=engine)

data_authors

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In [19]:
data_authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


In [20]:
# Menyusun kueri SQL untuk mendapatkan nama kolom dari tabel publishers
query_authors = """ 
SELECT column_name 
FROM information_schema.columns
WHERE table_name = 'authors';"""

# Membaca nama kolom ke dalam dataframe
data_authors = pd.io.sql.read_sql(query_authors, con=engine)
print("Nama kolom dalam tabel authors:")
data_authors


Nama kolom dalam tabel authors:


column_name
0   author_id
1      author

### Explore table books

In [21]:
# Menampilkan data di table authors
query_books = """SELECT * FROM public.books ;"""
data_books = pd.io.sql.read_sql(query_books, con=engine)
data_books


book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [22]:
data_books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


In [23]:
# Menyusun kueri SQL untuk mendapatkan nama kolom dari tabel publishers
query_books = """ 
SELECT column_name 
FROM information_schema.columns
WHERE table_name = 'books';"""

# Membaca nama kolom ke dalam dataframe
data_books = pd.io.sql.read_sql(query_books, con=engine)
print("Nama kolom dalam tabel books:")
data_books


Nama kolom dalam tabel books:


column_name
0           book_id
1         author_id
2         num_pages
3  publication_date
4      publisher_id
5             title

### Explore table ratings

In [24]:
# Menampilkan data di table authors
query_ratings = """SELECT * FROM public.ratings ;"""
data_ratings = pd.io.sql.read_sql(query_ratings, con=engine)
data_ratings


rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [25]:
data_ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


In [26]:
# Menyusun kueri SQL untuk mendapatkan nama kolom dari tabel publishers
query_ratings = """ 
SELECT column_name 
FROM information_schema.columns
WHERE table_name = 'ratings';"""

# Membaca nama kolom ke dalam dataframe
data_ratings = pd.io.sql.read_sql(query_ratings, con=engine)
print("Nama kolom dalam tabel ratings:")
data_ratings


Nama kolom dalam tabel ratings:


column_name
0   rating_id
1     book_id
2      rating
3    username

### Explore table reviews

In [27]:
# Menampilkan data di table authors
query_reviews = """SELECT * FROM public.reviews ;"""
data_reviews = pd.io.sql.read_sql(query_reviews, con=engine)
data_reviews


review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

In [28]:
data_reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


In [29]:
# Menyusun kueri SQL untuk mendapatkan nama kolom dari tabel publishers
query_reviews = """ 
SELECT column_name 
FROM information_schema.columns
WHERE table_name = 'reviews';"""

# Membaca nama kolom ke dalam dataframe
data_reviews = pd.io.sql.read_sql(query_reviews, con=engine)
print("Nama kolom dalam tabel reviews:")
data_reviews


Nama kolom dalam tabel reviews:


column_name
0   review_id
1     book_id
2    username
3        text

### Explore table lainnya

In [30]:
# Menampilkan data di table authors
query_orders = """SELECT * FROM public.orders ;"""
data_orders = pd.io.sql.read_sql(query_orders, con=engine)
data_orders


id               buyts revenue                   uid
0          1 2017-06-01 00:10:00  $17.00  10329302124590727494
1          2 2017-06-01 00:25:00   $0.55  11627257723692907447
2          3 2017-06-01 00:27:00   $0.37  17903680561304213844
3          4 2017-06-01 00:29:00   $0.55  16109239769442553005
4          5 2017-06-01 07:58:00   $0.37  14200605875248379450
...      ...                 ...     ...                   ...
50410  50411 2018-05-31 23:50:00   $4.64  12296626599487328624
50411  50412 2018-05-31 23:50:00   $5.80  11369640365507475976
50412  50413 2018-05-31 23:54:00   $0.30   1786462140797698849
50413  50414 2018-05-31 23:56:00   $3.67   3993697860786194247
50414  50415 2018-06-01 00:02:00   $3.42     83872787173869366

[50415 rows x 4 columns]

In [31]:
# Menyusun kueri SQL untuk mendapatkan nama kolom dari tabel publishers
query_orders = """ 
SELECT column_name 
FROM information_schema.columns
WHERE table_name = 'orders';"""

# Membaca nama kolom ke dalam dataframe
data_orders = pd.io.sql.read_sql(query_orders, con=engine)
print("Nama kolom dalam tabel orders:")
data_orders


Nama kolom dalam tabel orders:


column_name
0          id
1       buyts
2     revenue
3         uid

In [32]:
# Menampilkan data di table authors
query_visits = """SELECT * FROM public.visits ;"""
data_visits = pd.io.sql.read_sql(query_visits, con=engine)
data_visits


id                   uid   device               endts  sourceid  \
0            0  16879256277535980062    touch 2017-12-20 17:38:00         4   
1            1    104060357244891740  desktop 2018-02-19 17:21:00         2   
2            2   7459035603376831527    touch 2017-07-01 01:54:00         5   
3            3  16174680259334210214  desktop 2018-05-20 11:23:00         9   
4            4   9969694820036681168  desktop 2017-12-27 14:06:00         3   
...        ...                   ...      ...                 ...       ...   
358527  358527    272012551460639309  desktop 2018-03-20 15:25:00         4   
358528  358528  11171727274084046837    touch 2018-05-15 23:09:00         2   
358529  358529  13482259146982443019  desktop 2018-05-06 20:27:00         5   
358530  358530   3266363999948772866    touch 2018-04-03 11:23:00         5   
358531  358531   5583642936685644312  desktop 2018-02-12 20:32:00         3   

                   startts  
0      2017-12-20 17:20:00  
1      2018-02-19 16:53:00  
2      2017-07-01 01:54:00  
3      2018-05-20 10:59:00  
4      2017-12-27 14:06:00  
...                    ...  
358527 2018-03-20 15:11:00  
358528 2018-05-15 23:08:00  
358529 2018-05-06 20:27:00  
358530 2018-04-03 11:19:00  
358531 2018-02-12 20:31:00  

[358532 rows x 6 columns]

In [33]:
# Tugas Menyusun kueri SQL untuk mendapatkan nama kolom dari tabel publishers
query_visits = """ 
SELECT column_name 
FROM information_schema.columns
WHERE table_name = 'visits';"""

# Membaca nama kolom ke dalam dataframe
data_visits = pd.io.sql.read_sql(query_visits, con=engine)
print("Nama kolom dalam tabel visits:")
data_visits


Nama kolom dalam tabel visits:


column_name
0          id
1       endts
2    sourceid
3     startts
4         uid
5      device

## Tugas

### Temukan jumlah buku yang dirilis setelah 1 Januari 2000

In [34]:
# Temukan jumlah buku yang dirilis setelah 1 Januari 2000.

# Membuat query SQL
query_satu = """
SELECT count(*) AS jumlah_buku
FROM public.books
WHERE publication_date > '2000-01-01';
"""

# Membaca hasil query ke dalam dataframe
baca_query = pd.io.sql.read_sql(query_satu, con=engine)

# Menampilkan hasil
print(baca_query)

   jumlah_buku
0          819


Kesimpulan: 

Jumlah buku yang dirilis setelah 1 Januari 2000 adalah [819] buku. Informasi ini dapat digunakan untuk memahami sejauh mana buku-buku yang lebih baru berperan dalam pasar buku.

### Hitung jumlah ulasan pengguna dan rating rata-rata untuk setiap buku

In [35]:
# Tugas :
# ● Temukan jumlah ulasan pengguna dan rata-rata peringkat untuk setiap buku.  

# Membuat query ke-dua
query_dua = """
SELECT 
    b.book_id, 
    b.title AS judul_buku, 
    COUNT(r.review_id) AS jumlah_ulasan,
    AVG(rt.rating) AS rata_rata_peringkat 
FROM public.books as b
LEFT JOIN 
    public.reviews AS r ON b.book_id = r.book_id
LEFT JOIN 
    public.ratings AS rt ON b.book_id = rt.book_id
GROUP BY 
    b.book_id, title
ORDER BY
    jumlah_ulasan;
"""

# Membaca hasil kueri ke dalam dataframe
baca_query = pd.io.sql.read_sql(query_dua, con=engine)

# Menampilkan hasil
#print(df)
baca_query



book_id                                         judul_buku  \
0        387                               Leonardo's Notebooks   
1        221                          Essential Tales and Poems   
2        672   The Cat in the Hat and Other Dr. Seuss Favorites   
3         83    Anne Rice's The Vampire Lestat: A Graphic Novel   
4        808                            The Natural Way to Draw   
..       ...                                                ...   
995      299  Harry Potter and the Chamber of Secrets (Harry...   
996      302  Harry Potter and the Prisoner of Azkaban (Harr...   
997      673                             The Catcher in the Rye   
998      750                The Hobbit  or There and Back Again   
999      948                            Twilight (Twilight  #1)   

     jumlah_ulasan  rata_rata_peringkat  
0                0             4.000000  
1                0             4.000000  
2                0             5.000000  
3                0             3.666667  
4                0             3.000000  
..             ...                  ...  
995            480             4.287500  
996            492             4.414634  
997            516             3.825581  
998            528             4.125000  
999           1120             3.662500  

[1000 rows x 4 columns]

Kesimpulan: 

Kita telah menghitung jumlah ulasan pengguna dan rata-rata peringkat untuk setiap buku. 
Informasi ini dapat memberikan wawasan tentang popularitas dan kualitas buku-buku tertentu, yang dapat digunakan dalam pengambilan keputusan terkait dengan perpustakaan atau penjualan buku.

### Identifikasi penerbit yang telah menerbitkan jumlah buku terbanyak

In [36]:
#tugas :
#● Identifikasi penerbit yang telah merilis jumlah buku terbanyak dengan lebih dari 50 halaman 
#  (ini akan membantu Anda mengeluarkan brosur dan publikasi serupa dari analisis Anda).

# Membuat query ke-tiga
query_tiga = """
SELECT
    p.publisher_id,
    p.publisher AS nama_penerbit,
    COUNT(b.book_id) AS jumlah_buku_terbit
FROM
    public.publishers AS p
JOIN
    public.books AS b ON p.publisher_id = b.publisher_id
WHERE
    num_pages > 50
GROUP BY
    p.publisher_id, nama_penerbit
ORDER BY
    jumlah_buku_terbit DESC
LIMIT 10;
"""

# Membaca  hasil query kedalam dataframe
baca_query = pd.io.sql.read_sql(query_tiga, con=engine)

# Menampilkan hasil 
#print(baca_query)
baca_query

publisher_id             nama_penerbit  jumlah_buku_terbit
0           212             Penguin Books                  42
1           309                   Vintage                  31
2           116  Grand Central Publishing                  25
3           217          Penguin Classics                  24
4            33          Ballantine Books                  19
5            35                    Bantam                  19
6            45                   Berkley                  17
7           284        St. Martin's Press                  14
8            46             Berkley Books                  14
9            83                     Delta                  13

Kesimpulan: 
Data yang di tampilkan 10 raw teratas.

[Penguin Books] adalah penerbit yang telah merilis jumlah buku terbanyak dengan lebih dari 50 halaman. Informasi ini dapat membantu dalam mengenali penerbit yang memiliki kontribusi besar dalam industri buku dalam menerbitkan buku-buku.

### Identifikasi penulis dengan rata-rata peringkat buku tertinggi

In [37]:
# ● Identifikasi penulis dengan rata-rata peringkat buku tertinggi (lihat hanya buku dengan setidaknya 50 ulasan).

# Membuat query ke-empat
query_empat = """
SELECT
    a.author_id,
    a.author AS nama_penulis,
    AVG(rt.rating) AS rata_rata_peringkat,
    COUNT(*) AS jumlah_buku
FROM
    public.authors AS a
JOIN
    public.books AS b ON a.author_id = b.author_id
JOIN
    (
        SELECT book_id
        FROM public.ratings
        GROUP BY book_id
        HAVING COUNT(*) >= 50
    ) AS rated_books ON b.book_id = rated_books.book_id
JOIN
    public.ratings AS rt ON b.book_id = rt.book_id
GROUP BY
    a.author_id, nama_penulis
ORDER BY
    rata_rata_peringkat DESC
LIMIT 1;
"""

# Membaca hasil kueri ke dalam dataframe
baca_query = pd.io.sql.read_sql(query_empat, con=engine)

# Menampilkan hasil
baca_query




author_id                nama_penulis  rata_rata_peringkat  jumlah_buku
0        236  J.K. Rowling/Mary GrandPré             4.287097          310

Kesimpulan: 

Data yang ditampilkan tidak ada limitnya.

[J.K. Rowling/Mary GrandPré] adalah penulis dengan rata-rata peringkat buku tertinggi, memiliki rata-rata_peringkat [4.287097] dengan jumlah buku [310] ini adalah data bahwa hanya buku dengan setidaknya 50 ulasan yang dipertimbangkan. 

Informasi ini dapat digunakan untuk mengetahui kualitas tulisan penulis dalam menghasilkan buku-buku yang mendapatkan peringkat tinggi dari pembaca.

### Temukan rata-rata jumlah ulasan teks di antara pengguna yang memberi peringkat lebih dari 50 buku

In [38]:
#● Temukan rata-rata jumlah ulasan teks di antara pengguna yang memberi peringkat lebih dari 50 buku.

# membuat query_lima
query_lima = """
SELECT AVG(count) AS rata_rata_jumlah_ulasan_teks
FROM (
    SELECT COUNT(reviews.text) 
    FROM
        (SELECT username
        FROM ratings
        GROUP BY username
        HAVING COUNT(rating) > 50 )AS users
    JOIN reviews ON users.username = reviews.username
    GROUP BY users.username) AS count
    
"""

# Membaca hasil kueri ke dalam dataframe
baca_query = pd.io.sql.read_sql(query_lima, con=engine)

# Menampilkan hasil
#print(baca_query)
baca_query

rata_rata_jumlah_ulasan_teks
0                     24.333333

Kesimpulan: 

Rata-rata jumlah ulasan teks di antara pengguna yang memberi peringkat lebih dari 50 buku adalah [24.33]. 

Ini menggambarkan tingkat partisipasi dan keterlibatan pengguna dalam memberikan ulasan teks, 
Hal ini dapat berguna dalam memahami tingkat interaksi antara pengguna dengan buku-buku yang telah diterbitkan.

Kesimpulan umum:

**1. Jumlah Buku Setelah 1 Januari 2000:**

Pensortiran ditemukan ada sejumlah buku yang dirilis setelah 1 Januari 2000. Jumlahnya ditemukan dengan menghitung buku-buku yang memiliki tanggal publikasi setelah tanggal tersebut.
Jumlah buku yang dirilis setelah 1 Januari 2000 adalah [819] buku. Informasi ini dapat digunakan untuk memahami sejauh mana buku-buku yang lebih baru berperan dalam pasar buku.

**2. Jumlah Ulasan dan Rata-Rata Peringkat untuk Setiap Buku:**

Setiap buku memiliki jumlah ulasan pengguna dan rata-rata peringkat yang dapat digunakan untuk mengevaluasi popularitas dan kualitasnya.
Kita telah menghitung jumlah ulasan pengguna dan rata-rata peringkat untuk setiap buku. Informasi ini dapat memberikan wawasan tentang popularitas dan kualitas buku-buku tertentu, yang dapat digunakan dalam pengambilan keputusan terkait dengan perpustakaan atau penjualan buku.

**3. Penerbit dengan Banyak Buku Lebih dari 50 Halaman:**

Penerbit yang merilis banyak buku dengan lebih dari 50 halaman dapat diidentifikasi. Ini membantu dalam menyingkirkan publikasi yang tidak relevan yaitu buku dengan halaman < 50.
[Penguin Books] adalah penerbit yang telah merilis jumlah buku terbanyak dengan lebih dari 50 halaman. Informasi ini dapat membantu dalam mengenali penerbit yang memiliki kontribusi besar dalam industri buku dalam menerbitkan buku-buku.

**4. Penulis dengan Rata-Rata Peringkat Tertinggi:**

Penulis yang memiliki rata-rata peringkat tertinggi untuk buku-bukunya dapat diidentifikasi. Ini memberikan informasi tentang 
penulis yang sangat disukai oleh pembaca.
[J.K. Rowling/Mary GrandPré] adalah penulis dengan rata-rata peringkat buku tertinggi, memiliki rata-rata_peringkat [4.287097] dengan jumlah buku [310] ini adalah data bahwa hanya buku dengan setidaknya 50 ulasan yang dipertimbangkan.

Informasi ini dapat digunakan untuk mengetahui kualitas tulisan penulis dalam menghasilkan buku-buku yang mendapatkan peringkat tinggi dari pembaca.

**5. Rata-Rata Jumlah Ulasan Teks dari Pengguna yang Memberi Peringkat Lebih dari 50 Buku:**

Rata-rata jumlah ulasan teks dari pengguna yang memberi peringkat pada lebih dari 50 buku dapat digunakan untuk mengukur partisipasi dan kontribusi pengguna yang aktif dalam memberikan ulasan buku.
Rata-rata jumlah ulasan teks di antara pengguna yang memberi peringkat lebih dari 50 buku adalah [24.33].

Ini menggambarkan tingkat partisipasi dan keterlibatan pengguna dalam memberikan ulasan teks, Hal ini dapat berguna dalam memahami tingkat interaksi antara pengguna dengan buku-buku yang telah diterbitkan.